src/alloc_pool.c

```diff
@@ -44,20 +46,97 @@ DEQ_DECLARE(qd_alloc_type_t, qd_alloc_type_list_t);
 #define PATTERN_BACK  0xbabecafe
 
 struct qd_alloc_item_t {
-    DEQ_LINKS(qd_alloc_item_t);
     uint32_t              sequence;
 #ifdef QD_MEMORY_DEBUG
     qd_alloc_type_desc_t *desc;
     uint32_t              header;
 #endif
 };
 
-DEQ_DECLARE(qd_alloc_item_t, qd_alloc_item_list_t);
+#define CHUNK_SIZE 128/sizeof(void*)
 
+struct qd_alloc_chunk_t {
+    qd_alloc_chunk_t     *prev;
+    qd_alloc_item_t      *items[CHUNK_SIZE];
+    qd_alloc_chunk_t     *next;
+};
+
+struct qd_alloc_linked_stack_t {
+    //the base
+    qd_alloc_chunk_t     *top_chunk;
+    uint32_t              top;          //qd_alloc_item* top_item = top_chunk->items[top+1] <-> top > 0
```

Comment:
Probably want to reorder `top` and `size`. On 64 bit machine there would be 32 bits of padding in the middle of the struct with this ordering.

Reply:
Actually as `qp_alloc_chunk_t` contains only pointers you probably want to put `size` as the very last member in this structure (but you'll still get some padding)

Reply:
We can try to reorder fields and see if it will get better cache localty: I see that `qd_alloc_linked_stack_t` is suboptimal, but I've just tried to put `base_chunk` near to the most accessed fields to make used of the adiacent line prefetching of many Intel CPUs; so I've focused just on that...
I can see too that `CAPACITY` could be changed to improve things too...

Reply:
In this case I was just thinking about eliminating useless struct padding - which might allow you to fit more into a cache line, but might not help locality or prefetching.

Reply:
I'm not aware of any struct padding here...but maybe there are some introduced automagically somehwere?

Reply:
On a 64 bit machine you'll get padding after the uint32 (`top`) to pad it out to 64 bits. Everything else in the struct is 64 bits (pointers) so to avoid the padding in the middle of the struct you want to put `top` at the end of the struct.
[On a 32 bit machine this is not a concern, but don't care as much about them anyway].

Reply:
> to avoid the padding in the middle of the struct you want to put top at the end of the struct.

I haven't considered field padding, you're right!
In this case, considering that `base_chunk` is > 128 bytes I won't risk to get a cache miss on each access?

Reply:
Hmm, difficult to say - in any event it's going to be close to some entries in the list and far from others. Perhaps eliminating it completely might be possible.

Another thing to think about is not to use pointers at all, but rather use a base (64 bit address) and smaller offset (32 bit entry count) - that should allow more entries into the same number of cache lines - you would have to allocate larger and contiguous blocks though.

Reply:
I love the "compressed pointers" idea and given that we are mostly memory limited here it would allow to pack more pointers...something that using an intrusive pointer solution was really tricky (if not impossible) to implemente...